internal/backends/sqlite/backend.go

```diff
@@ -29,37 +31,79 @@ import (
 // backend implements backends.Backend interface.
 type backend struct {
 	r   *metadata.Registry
-	dir string
+	uri *url.URL
 }
 
 // NewBackendParams represents the parameters of NewBackend function.
 type NewBackendParams struct {
-	Dir string
+	URI string
 	L   *zap.Logger
 }
 
 // NewBackend creates a new SQLite backend.
 func NewBackend(params *NewBackendParams) (backends.Backend, error) {
-	fi, err := os.Stat(params.Dir)
+	uri, err := validateURI(params.URI)
 	if err != nil {
-		return nil, lazyerrors.Errorf("%q should be an existing directory: %w", params.Dir, err)
+		return nil, err
 	}
 
-	if !fi.IsDir() {
-		return nil, lazyerrors.Errorf("%q should be an existing directory", params.Dir)
-	}
-
-	r, err := metadata.NewRegistry(params.Dir, params.L.Named("metadata"))
+	r, err := metadata.NewRegistry(uri, params.L.Named("metadata"))
 	if err != nil {
 		return nil, lazyerrors.Error(err)
 	}
 
 	return backends.BackendContract(&backend{
 		r:   r,
-		dir: params.Dir,
+		uri: uri,
 	}), nil
 }
 
+// validateURI checks given URI value and returns parsed URI.
+// URI should contain 'file' scheme and point to an existing directory.
+// Path should end with '/'. Authority should be empty.
+// If URI contains local path it would be set to Path.
+func validateURI(value string) (*url.URL, error) {
+	uri, err := url.Parse(value)
+	if err != nil {
+		return nil, lazyerrors.Errorf("failed to parse backend URI: %w", err)
+	}
+
+	if uri.Scheme != "file" {
+		return nil, lazyerrors.Errorf("backend URI should have file scheme: %q", value)
+	}
+
+	if uri.User != nil {
+		return nil, lazyerrors.Errorf("backend URI should not contain user: %q", value)
+	}
+
+	if uri.Host != "" {
+		return nil, lazyerrors.Errorf("backend URI should not contain host: %q", value)
+	}
+
+	dir := uri.Path
+
+	if uri.Opaque != "" {
+		dir = uri.Opaque
```

Comment:
What condition is it not sufficient to check `uri.Path` contains `/` but must check `uri.Opaque`? Is it something to do with query parameter? :thinking:

Reply:
It’s how `url.URL` works. Local paths would be in the `Opaque` field.